### PR TITLE
Taha

### PR DESCRIPTION
**Pull request by tazzaoui**

### CODE REVIEW

File: cortix/src/network.py [hunk 1]
```diff
@@ -225,60 +227,37 @@ def run(self):
...
-                    self.modules[mod_idx]))
-
             # Synchronize at the end
             for p in processes:
                 p.join()
 
-        return
+        # Reload saved modules
+        for mod in self.modules:
+            self.modules = list()
```
[the remark this answers is not in the document]
Comment (tazzaoui) [reply cortix/src/network.py]: Good catch. I just posted a fix for this in `network.py` that avoids looping over a changing list on line 255:

~~~python
        for mod in self.modules:
            for file_name in os.listdir(".ctx-saved"):
                if file_name.endswith(".pkl"):
                    file_name = os.path.join(".ctx-saved", file_name)
                    with open(file_name, "rb") as f:
                        old_mods.append(pickle.load(f))
        self.modules = old_mods
~~~

[the remark this answers is not in the document]
Comment (dealmeidavf) [reply cortix/src/network.py]: All good.

File: cortix/src/network.py [hunk 2]
```diff
@@ -225,60 +227,38 @@ def run(self):
             if self.rank != 0:
                 mod = self.modules[self.rank-1]
                 self.log.info('Launching Module {}'.format(mod))
-                mod.run()
+                mod.run_and_save()
 
             # Sync here at the end
             self.comm.Barrier()
 
-            if len(self.modules) > self.max_n_modules_for_data_copy_on_root:
```
Comment (dealmeidavf) [inline cortix/src/network.py]: Could you leave the in-memory copy for now as an alternative and also so the examples and notebook are not broken? I don't have time to fix them now and it does not hurt to have both options: in-core and out-of-core saves.
If you run the examples and the notebook, they should all pass as they are at the moment after you added the out-of-core save. 
Thanks.

**Comment by tazzaoui** [comment]
> Is it ready to merge? Will the notebooks run?

Yes, everything is good to go. The notebooks need updating. Will create a pull request in cortix-nb.

**Comment by dealmeidavf** [comment]
> > Is it ready to merge? Will the notebooks run?
> 
> Yes, everything is good to go. The notebooks need updating. Will create a pull request in cortix-nb.

Ok. I will merge this PR and the PR for the notebooks. After that, please update PyPI.
Thanks.

**Pull request merged**